staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -674,7 +675,221 @@ type crdConversionRESTOptionsGetter struct {
+			dropInvalidMetadata: true,
+		}}
+		c := validatingConverter{delegate: t.converter, validator: unstructuredSchemaValidator{}}
+		ret.StorageConfig.Codec = versioning.NewCodec(ret.StorageConfig.Codec, d, c, &unstructuredCreator{}, discovery.NewUnstructuredObjectTyper(), &unstructuredDefaulter{delegate: Scheme}, t.encoderVersion, t.decoderVersion)
 	}
 	return ret, err
 }
+
+// validatingDecoder calls the delegate decoder, and then applies the Unstructured validator.
+type validatingDecoder struct {
+	delegate  runtime.Decoder
+	validator unstructuredSchemaValidator
+}
+
+var _ runtime.Decoder = validatingDecoder{}
+
+func (d validatingDecoder) Decode(data []byte, defaults *schema.GroupVersionKind, into runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
+	obj, gvk, err := d.delegate.Decode(data, defaults, into)
+	if err != nil {
+		return nil, gvk, err
+	}
+	if u, ok := obj.(*unstructured.Unstructured); ok {
+		if err := d.validator.apply(u); err != nil {
+			return nil, gvk, err
+		}
+	}
+
+	return obj, gvk, nil
+}
+
+// validatingConverter calls the delegate converter and applies the Unstructured validator.
+type validatingConverter struct {
+	delegate  runtime.ObjectConvertor
+	validator unstructuredSchemaValidator
+}
+
+var _ runtime.ObjectConvertor = validatingConverter{}
+
+func (v validatingConverter) Convert(in, out, context interface{}) error {
+	if err := v.delegate.Convert(in, out, context); err != nil {
+		return err
+	}
+
+	if u, ok := out.(*unstructured.Unstructured); ok {
+		if err := v.validator.apply(u); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (v validatingConverter) ConvertToVersion(in runtime.Object, gv runtime.GroupVersioner) (runtime.Object, error) {
+	out, err := v.delegate.ConvertToVersion(in, gv)
+	if err != nil {
+		return nil, err
+	}
+
+	if u, ok := out.(*unstructured.Unstructured); ok {
+		if err := v.validator.apply(u); err != nil {
+			return nil, err
+		}
+	}
+
+	return out, nil
+}
+
+func (v validatingConverter) ConvertFieldLabel(version, kind, label, value string) (string, string, error) {
+	return v.ConvertFieldLabel(version, kind, label, value)
+}
+
+// unstructuredSchemaValidator does the validation for Unstructured that json.Unmarshal
+// does for native types. This includes:
+// - validating and pruning ObjectMeta (here with optional error instead of pruning)
+// - TODO: application of an OpenAPI validator (against the whole object or a top-level field of it).
+// - TODO: optionally application of post-validation algorithms like defaulting and/or OpenAPI based pruning.
+type unstructuredSchemaValidator struct {
+	dropInvalidMetadata bool
+}
+
+func (v *unstructuredSchemaValidator) apply(u *unstructured.Unstructured) error {
+	// save implicit meta fields that don't have to be specified in the validation spec
+	kind, _, err := unstructured.NestedString(u.UnstructuredContent(), "kind")
```

Comment:
this ignores whether kind was set or not... if there was no kind field, this results in setting "kind"="". should we preserve that info and avoid calling SetKind if it wasn't set originally? same applies for apiVersion

Reply:
fixed